### DISCUSSION TITLE
Best practices for curvature optimization?

### DISCUSSION THREAD

**Discussion by gatoniel** [geoopt/geoopt#167] [Q&A]
Hi,

I use the `PoincareBall` for an embedding of data points. When I fix curvature at -1 the embedding works an generates a reasonable final loss. But when I fit also curvature it rapidly tends towards 0 while the loss stays above the optimal loss for fixed curvature.

Do you have any tips on how to fit curvature, e.g., using a different learning rate for this parameter?

**Comment by ferrine** [comment]
My best practice to optimize curvature is to have an unconstrained parameter in the Tangent space of zero and use exp map

```python
tens = torch.nn.Parameter(torch.zeros(10))
ball = geoopt.Stereographic(learnable=True)
opt = torch.optim.Adam([tens, *ball.parameters()])
for i in range(nsteps):
    point  = ball.expmap0(tens)
    ...
    opt.step()
```
This worked the best for me (long time ago). 

I also tried to use 2 optimizers as I had concerns about joint optimization.
```python
tens = torch.nn.Parameter(torch.zeros(10))
ball = geoopt.Stereographic(learnable=True)
opt1 = torch.optim.Adam([tens])
opt2 = torch.optim.Adam(ball.parameters())
for i in range(nsteps):
    point  = ball.expmap0(tens)
    ...
    if i % 2 == 0:
        opt1.step()
    else:
        opt2.step()
```
